File: pythonlibrary/aibrd/generators/rtm.py
from datetime import date
from ..models.outputs import BRDContent


def generate_rtm(contents: list[BRDContent], project_name: str) -> str:
    today = date.today().isoformat()
    lines = [
        "# index.md — Requirement Traceability Matrix",
        f"_Project: {project_name} | Updated: {today}_",
        "",
    ]

    modules = [c.module_slug for c in contents if c.module_slug]
    if modules:
        lines += ["## Modules", ""]
        for slug in modules:
            lines.append(f"- [{slug}](modules/{slug}/CONTEXT.md)")
        lines.append("")

    lines += ["## Traceability Matrix", ""]
    lines.append("| ID | Requirement | Test Cases | Status |")
    lines.append("|---|---|---|---|")

    for content in contents:
        for flow in content.flows:
            tc_ids = [ac.id.replace("AC", "TC") for ac in content.criteria if ac.related_flow == flow.id]
            status = "✅" if tc_ids else "❌"
            tests = ", ".join(tc_ids) if tc_ids else "—"
            lines.append(f"| {flow.id} | {flow.name} | {tests} | {status} |")

        for rule in content.rules:
            lines.append(f"| {rule.id} | {rule.description[:60]} | — | ❌ |")

    lines += ["", "_Status: ✅ Covered · ⚠️ Partial · ❌ Missing_"]
    return "\n".join(lines)

File: vscodebase/src/core/generators/rtm.ts
import { BRDContent, RTMEntry } from '../models/outputs'

export function generateRTM(
  contents: BRDContent[],
  projectName: string
): string {
  const entries = buildRTMEntries(contents)
  const now = new Date().toISOString().split('T')[0]

  const lines: string[] = [
    '# index.md — Requirement Traceability Matrix',
    `_Project: ${projectName} | Updated: ${now}_`,
    ''
  ]

  const modules = [...new Set(contents.map(c => c.moduleSlug).filter(Boolean))]
  if (modules.length > 0) {
    lines.push('## Modules', '')
    for (const slug of modules) {
      lines.push(`- [${slug}](modules/${slug}/CONTEXT.md)`)
    }
    lines.push('')
  }

  lines.push('## Traceability Matrix', '')
  lines.push('| ID | Requirement | Test Cases | Status |')
  lines.push('|---|---|---|---|')

  for (const entry of entries) {
    const tests = entry.testCaseIds.length > 0 ? entry.testCaseIds.join(', ') : '—'
    const status = entry.status === 'covered' ? '✅' : entry.status === 'partial' ? '⚠️' : '❌'
    lines.push(`| ${entry.requirementId} | ${entry.requirementName} | ${tests} | ${status} |`)
  }

  lines.push('')
  lines.push('_Status: ✅ Covered · ⚠️ Partial · ❌ Missing_')

  return lines.join('\n')
}

function buildRTMEntries(contents: BRDContent[]): RTMEntry[] {
  const entries: RTMEntry[] = []

  for (const content of contents) {
    for (const flow of content.flows) {
      const relatedTC = content.criteria
        .filter(ac => ac.relatedFlow === flow.id)
        .map(ac => ac.id.replace('AC', 'TC'))
      entries.push({
        requirementId: flow.id,
        requirementName: flow.name,
        testCaseIds: relatedTC,
        codeFiles: [],
        status: relatedTC.length > 0 ? 'covered' : 'missing'
      })
    }

    for (const rule of content.rules) {
      entries.push({
        requirementId: rule.id,
        requirementName: rule.description.slice(0, 60),
        testCaseIds: [],
        codeFiles: [],
        status: 'missing'
      })
    }
  }

  return entries
}
